diff --git a/source/modules/lilv/lilv-0.21.3/lilv/lilvmm.hpp b/source/modules/lilv/lilv-0.21.3/lilv/lilvmm.hpp
index c3aa822..1f67e81 100644
--- a/source/modules/lilv/lilv-0.21.3/lilv/lilvmm.hpp
+++ b/source/modules/lilv/lilv-0.21.3/lilv/lilvmm.hpp
@@ -29,11 +29,6 @@
 
 namespace Lilv {
 
-static inline const char*
-uri_to_path(const char* uri) {
-	return lilv_uri_to_path(uri);
-}
-
 #define LILV_WRAP0(RT, prefix, name) \
 	inline RT name() { return lilv_ ## prefix ## _ ## name (me); }
 
@@ -329,6 +324,9 @@ struct World {
 	inline LilvNode* new_uri(const char* uri) const {
 		return lilv_new_uri(me, uri);
 	}
+	inline LilvNode* new_file_uri(const char* host, const char* path) const {
+		return lilv_new_file_uri(me, host, path);
+	}
 	inline LilvNode* new_string(const char* str) const {
 		return lilv_new_string(me, str);
 	}
