diff -U 3 -H -b -B -d -r -N -- lilv-0.14.4.old/src/util.c lilv-0.14.4/src/util.c
--- lilv-0.14.4.old/src/util.c	2012-08-09 21:51:00.000000000 +0100
+++ lilv-0.14.4/src/util.c	2012-09-15 01:20:07.908701939 +0100
@@ -29,7 +29,6 @@
 #include <string.h>
 
 #ifdef _WIN32
-#    define _WIN32_WINNT 0x0600  /* for CreateSymbolicLink */
 #    include <windows.h>
 #    include <direct.h>
 #    include <io.h>
@@ -426,7 +425,7 @@
 	int ret = 0;
 	if (strcmp(oldpath, newpath)) {
 #ifdef _WIN32
-		ret = !CreateSymbolicLink(newpath, oldpath, 0);
+		ret = 0;
 #else
 		ret = symlink(oldpath, newpath);
 #endif
