diff -U 3 -H -b -B -d -r -N -- lilv-0.14.4.old/lilv/lilvmm.hpp lilv-0.14.4/lilv/lilvmm.hpp
--- lilv-0.14.4.old/lilv/lilvmm.hpp	2012-09-13 12:47:55.000000000 +0100
+++ lilv-0.14.4/lilv/lilvmm.hpp	2012-09-13 12:48:10.950555311 +0100
@@ -60,6 +60,7 @@
 #endif
 
 struct Node {
+    inline Node(LilvNode* node)       : me(node) {}
 	inline Node(const LilvNode* node) : me(lilv_node_duplicate(node)) {}
 	inline Node(const Node& copy)     : me(lilv_node_duplicate(copy.me)) {}
 
