--- linuxsampler-static-1.0.0+svn2356.orig/configure.ac
+++ linuxsampler-static-1.0.0+svn2356/configure.ac
@@ -556,21 +556,6 @@ if test "x$MAC_PLUGIN_INSTALL_DIR" = "x"
 fi
 AC_SUBST(MAC_PLUGIN_INSTALL_DIR)
 
-# have we found at least one MIDI input and one audio output driver ?
-if test "$have_midi_input_driver" = "false"; then
-    echo "No supported MIDI input system found!"
-    echo "Sorry, LinuxSampler only supports the following MIDI drivers at the moment:"
-    echo "ALSA, JACK, MIDIShare, CoreMIDI, MME."
-    echo "If you think you have one of those available on your system, make sure you"
-    echo "also have the respective development (header) files installed."
-    exit -1;
-fi
-if test "$have_audio_output_driver" = "false"; then
-    echo "No supported audio output system found!"
-    echo "Sorry, LinuxSampler only supports ALSA, JACK, ARTS and ASIO as audio output"
-    echo "driver at the moment!"
-    exit -1;
-fi
 
 
 
