diff --git a/source/modules/lilv/sratom-0.4.7/sratom/sratom.h b/source/modules/lilv/sratom-0.4.7/sratom/sratom.h
index a86f67e..cf33396 100644
--- a/source/modules/lilv/sratom-0.4.7/sratom/sratom.h
+++ b/source/modules/lilv/sratom-0.4.7/sratom/sratom.h
@@ -23,9 +23,9 @@
 
 #include <stdint.h>
 
-#include "lv2/lv2plug.in/ns/ext/urid/urid.h"
-#include "lv2/lv2plug.in/ns/ext/atom/atom.h"
-#include "lv2/lv2plug.in/ns/ext/atom/forge.h"
+#include "lv2/urid.h"
+#include "lv2/atom.h"
+#include "lv2/atom-forge.h"
 #include "serd/serd.h"
 #include "sord/sord.h"
 
