diff --git a/source/modules/lilv/sratom-0.4.7/src/sratom.c b/source/modules/lilv/sratom-0.4.7/src/sratom.c
index e0996c4..b6ea4b1 100644
--- a/source/modules/lilv/sratom-0.4.7/src/sratom.c
+++ b/source/modules/lilv/sratom-0.4.7/src/sratom.c
@@ -20,9 +20,9 @@
 #include <stdlib.h>
 #include <string.h>
 
-#include "lv2/lv2plug.in/ns/ext/atom/forge.h"
-#include "lv2/lv2plug.in/ns/ext/atom/util.h"
-#include "lv2/lv2plug.in/ns/ext/midi/midi.h"
+#include "lv2/atom-forge.h"
+#include "lv2/atom-util.h"
+#include "lv2/midi.h"
 
 #include "sratom/sratom.h"
 
